plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeLocalConcurrentCreateTableTest.java

```diff
@@ -0,0 +1,182 @@
+            assertUpdate("DROP TABLE IF EXISTS test_ctas_3");
+            executor.shutdownNow();
+            assertThat(executor.awaitTermination(10, SECONDS)).isTrue();
+        }
+    }
+
+    @Test
+    public void testConcurrentCreateTableAsSelectSameTable()
+            throws Exception
+    {
+        testConcurrentCreateTableAsSelectSameTable(false);
+        testConcurrentCreateTableAsSelectSameTable(true);
+    }
+
+    private void testConcurrentCreateTableAsSelectSameTable(boolean partitioned)
+            throws Exception
+    {
+        int threads = 3;
+        CyclicBarrier barrier = new CyclicBarrier(threads);
+        ExecutorService executor = newFixedThreadPool(threads);
+        String tableName = "test_concurrent_ctas_" + randomNameSuffix();
+
+        try {
+            getQueryRunner().execute("create table stg_test as SELECT a, b, 20220101 as d FROM UNNEST(SEQUENCE(1, 9001), SEQUENCE(1, 9001)) AS t(a, b)");
+
+            String selectString = " as (select stg1.a as a, stg1.b as b, stg1.d as part from stg_test stg1, stg_test stg2 where stg1.d=stg2.d)";
+
+            // Execute concurrent CTAS operations
+            executor.invokeAll(IntStream.range(0, threads).mapToObj(i -> (Callable<Void>) () -> {
+                barrier.await(10, SECONDS);
+                try {
+                    getQueryRunner().execute("CREATE TABLE " + tableName +
+                            (partitioned ? " WITH (partitioned_by = ARRAY['part'])" : "") + selectString);
+                }
+                catch (Exception e) {
+                    RuntimeException trinoException = getTrinoExceptionCause(e);
+                    assertThat(trinoException).hasMessageContaining("Table already exists");
+                }
+                return null;
+            }).collect(toImmutableList())).forEach(MoreFutures::getDone);
+
+            // Verify table exists and has one row
+            assertThat((long) computeScalar("SELECT count(*) FROM " + tableName)).isEqualTo(81018001L);
+
+            // Verify table history shows single creation
+            assertQuery(
+                    "SELECT version, operation, isolation_level, is_blind_append FROM \"" + tableName + "$history\"",
+                    "VALUES (0, 'CREATE TABLE AS SELECT', 'WriteSerializable', true)");
+
+            // Ensure the files in the table directory contains only the file written by succeeded query
+            long pathCount = (long) computeScalar("select count(distinct(\"$path\")) as path FROM " + tableName);
+            long pathMatchingSuccessfulQueryCount = (long) computeScalar("with a as (select distinct(\"$path\") as path from " + tableName + "), " +
+                    "b as (select element_at(operation_parameters, 'queryId') as queryId from \"" + tableName + "$history\") " +
+                    "select count(1) from a,b where a.path like '%' || b.queryId || '%'");
+            assertThat(pathCount).isEqualTo(pathMatchingSuccessfulQueryCount);
```

Comment:
This assert does not check files existence. $history and $path will return only files from succeeded queries.
You can use `getTableFiles` like here:
https://github.com/trinodb/trino/blob/fbb961083b66f900ae0daabb8631fd482a7cc8e3/plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeConnectorSmokeTest.java#L671